Add deprecation warning on use of qiskit.IBMQ

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

With the pending release of the qiskit-ibm-provider that superscedes the
qiskit-ibmq-provider package. The qiskit-ibmq-provider is now deprecated
and will be retired/archived in the future. In coordination with that
this commit deprecates the lazy loading fixture qiskit.IBMQ and removes
documentation references to qiskit-ibmq-provider in the repo.

### Details and comments

This is on hold until the qiskit-ibm-provider package has its first release

garrison commented on qiskit/tools/monitor/overview.py:
```diff
@@ -59,6 +68,15 @@ def backend_monitor(backend):
+        "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+        "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+        "in a future release. Instead you should use the qiskit-ibm-provider package "
+        "which is accesible from 'qiskit_ibm_provider'.",
```
```suggestion
        "which is accessible from 'qiskit_ibm_provider'.",
```

garrison commented on qiskit/tools/monitor/overview.py:
```diff
@@ -27,6 +28,14 @@ def get_unique_backends():
+        "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+        "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+        "in a future release. Instead you should use the qiskit-ibm-provider package "
+        "which is accesible from 'qiskit_ibm_provider'.",
```
```suggestion
        "which is accessible from 'qiskit_ibm_provider'.",
```

garrison commented on qiskit/__init__.py:
```diff
@@ -122,6 +131,15 @@ def __getattr__(self, attr):
+                    "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+                    "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+                    "in a future release. Instead you should use the qiskit-ibm-provider package "
+                    "which is accesible from 'qiskit_ibm_provider'. Just replace 'qiskit.IBMQ' "
```
```suggestion
                    "which is accessible from 'qiskit_ibm_provider'. Just replace 'qiskit.IBMQ' "
```

garrison commented on qiskit/__init__.py:
```diff
@@ -112,6 +112,15 @@ def __bool__(self):
+                    "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+                    "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+                    "in a future release. Instead you should use the qiskit-ibm-provider package "
+                    "which is accesible from 'qiskit_ibm_provider'.",
```
```suggestion
                    "which is accessible from 'qiskit_ibm_provider'.",
```

HuangJunye commented on qiskit/__init__.py:
```diff
@@ -122,6 +131,15 @@ def __getattr__(self, attr):
+                    "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+                    "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+                    "in a future release. Instead you should use the qiskit-ibm-provider package "
+                    "which is accesible from 'qiskit_ibm_provider'. Just replace 'qiskit.IBMQ' "
+                    "with 'qiskit_ibm_provider.IBMProvider'",
```
This part has an additional sentence "Just replace 'qiskit.IBMQ' with 'qiskit_ibm_provider.IBMProvider'" Is this intended?

HuangJunye commented on qiskit/__init__.py:
```diff
@@ -112,6 +112,15 @@ def __bool__(self):
                 from qiskit.providers import ibmq
 
                 self.ibmq = ibmq.IBMQ
+                warnings.warn(
+                    "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+                    "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+                    "in a future release. Instead you should use the qiskit-ibm-provider package "
+                    "which is accesible from 'qiskit_ibm_provider'.",
```
```suggestion
                warnings.warn(
                    "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
                    "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
                    "in a future release. Instead you should use the qiskit-ibm-provider package "
                    "which is accesible from 'qiskit_ibm_provider'. You can install it with 'pip install qiskit_ibm_provider'",
```

HuangJunye replied:
Similarly, I suggest to add a sentence about pip install for all other deprecation warnings.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit-terra/pull/8080/commits/da1b5f0f9e410b620f16e52c04b4979d71861163

HuangJunye reviewed (changes requested):
Overall looks good to me. I did a quick search in the whole repo, and found a few places still need update the new import path. Can you please make those changes?

- https://github.com/Qiskit/qiskit-terra/blob/3241978340f534b76e4cf0e5e25ed16e2af5cf78/tools/update_fake_backends.py#L22
- https://github.com/Qiskit/qiskit-terra/blob/95a97ec53caddc885d02fd353e8d0c745328fd39/qiskit/tools/jupyter/__init__.py#L28
- https://github.com/Qiskit/qiskit-terra/blob/95a97ec53caddc885d02fd353e8d0c745328fd39/qiskit/tools/jupyter/__init__.py#L42
- https://github.com/Qiskit/qiskit-terra/blob/d3243e9e1d0e916e1d2c4af442011605204598b0/qiskit/providers/fake_provider/__init__.py#L73

I am not sure whether we should also update the path in this line in `ibmq_mock.py`, 
- https://github.com/Qiskit/qiskit-terra/blob/95a97ec53caddc885d02fd353e8d0c745328fd39/qiskit/test/ibmq_mock.py#L45

Do you need to add a test for deprecation and a release note as well?

mtreinish commented:
> Overall looks good to me. I did a quick search in the whole repo, and found a few places still need update the new import path. Can you please make those changes?
> 
>     * https://github.com/Qiskit/qiskit-terra/blob/3241978340f534b76e4cf0e5e25ed16e2af5cf78/tools/update_fake_backends.py#L22
> 
>     * https://github.com/Qiskit/qiskit-terra/blob/95a97ec53caddc885d02fd353e8d0c745328fd39/qiskit/tools/jupyter/__init__.py#L28
> 
>     * https://github.com/Qiskit/qiskit-terra/blob/95a97ec53caddc885d02fd353e8d0c745328fd39/qiskit/tools/jupyter/__init__.py#L42
> 
>     * https://github.com/Qiskit/qiskit-terra/blob/d3243e9e1d0e916e1d2c4af442011605204598b0/qiskit/providers/fake_provider/__init__.py#L73
> 
> 

Done in: https://github.com/Qiskit/qiskit-terra/pull/8080/commits/e4ed253575c93da0c5ec79c1730e6a65bbdb8afd

> I am not sure whether we should also update the path in this line in `ibmq_mock.py`,

I just dropped that module in: https://github.com/Qiskit/qiskit-terra/pull/8080/commits/e4ed253575c93da0c5ec79c1730e6a65bbdb8afd it was only there for legacy reasons in the docs. It's sole usage was for jupyter execute docs to be able to use `qiskit.IBMQ` and mock the calls to return a fake backend. Since we don't care about pretending that the ibm provider is part of the qiskit namespace anymore and we're no longer using jupyter execute this had no relevance so it was easier to just remove it as I don't think anything externally should have been using it.

> 
>     * https://github.com/Qiskit/qiskit-terra/blob/95a97ec53caddc885d02fd353e8d0c745328fd39/qiskit/test/ibmq_mock.py#L45
> 
> 
> Do you need to add a test for deprecation and a release note as well?

I added a release note in https://github.com/Qiskit/qiskit-terra/pull/8080/commits/7ec28f4e158c6b6dd274001d6eb09c568d3d37b3 tests for this are a bit weird as they require the ibmq provider to be installed which we purposely don't install in tests. All we'd be testing is that a deprecation warning is emitted as we can't really test anything functional about the alias without having credentials and ibmq installed.

mtreinish commented:
>  as I don't think anything externally should have been using it.

Sigh, I stand corrected the ibmq provider was using it for it's jupyter-execute docs which were copied over from terra when that split out (because of course why would it be simple): https://github.com/Qiskit/qiskit-ibmq-provider/search?q=mock_get_backend let me add the function back in for now. We can just drop it when we drop `qiskit.IBMQ`

HuangJunye commented on releasenotes/notes/deprecate-qiskit-ibmq-f0dc372526fe0c57.yaml:
```diff
@@ -0,0 +1,11 @@
+    maintaining this alias is no longer relevant with the new package. If you
+    were relying on the ``qiskit.IBMQ`` alias you should update your usage
+    to use ``qiskit.providers.ibmq.IBMQ`` directly instead (and also consider
+    migrating to ``qiskit-ibm-provider``).
```
Does qiskit-ibm-provider has its own documentation site? Can we link to the migration guide (either the [source file](https://github.com/Qiskit/qiskit-ibm-provider/blob/main/docs/tutorials/Migration_Guide_from_qiskit-ibmq-provider.ipynb) or the rendered page on the documentation site) here?

HuangJunye replied:
Found the link to the migration guide on the doc site: https://qiskit.org/documentation/partners/qiskit_ibm_provider/tutorials/Migration_Guide_from_qiskit-ibmq-provider.html

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit-terra/pull/8080/commits/c2d21b435435cc097186e684c78bbecf1004e281

HuangJunye reviewed (commented):
Just one small comment for the release note.

HuangJunye reviewed (approved):
LGTM!
